### ISSUE TITLE
v2.51.3 is overdue

### ISSUE THREAD

**Issue by lilyball** [bcpierce00/unison#245]
v2.51.2 is over a year old now. Can you please put out a new release? v2.51.2 does not include #164, which is blocking the ability to update the `unison` package in Nix (https://github.com/NixOS/nixpkgs/issues/50255).

**Comment by pascalandy** [comment]
It also great to push releases as more and more people watch `Release Only` :)

**Comment by pascalandy** [comment]
Well ?

**Comment by lilyball** [comment]
Oh oops, I was doing some cleanup of my old issues and misread this one, this is still an open issue.

**Comment by aadrian** [comment]
Any news on this? or is the project abandoned?

Thank you.

**Comment by bcpierce00** [comment]
The project is not abandoned, but I am aware that I am late putting together a new release.  (Excuses: First the pandemic ate all my time, then I decided it was important to put a lot of energy toward trying to build a really good platform <http://clowdr.org/> for virtual academic conferences…)  

If someone has the cycles and GitHub / OCaml expertise to help do this, I’d be most grateful…

   - Benjamin

**Comment by glondu** [comment]
> If someone has the cycles and GitHub / OCaml expertise to help do this, I’d be most grateful…

I do, and I do have Linux expertise (I am the Debian maintainer of Unison), but I do not have Windows nor Mac OS X expertise. How do you propose we proceed?

**Comment by bcpierce00** [comment]
I was hoping you would be the one to respond, Stéphane!  :-)  Many thanks for the offer of help, which I gratefully accept.

I've invited you to the repo.  Please take a look at pending issues / PRs (including yours!) and see if there are any you want to address before tagging a new release.

In general, we try to avoid making changes that require changing the major version number, to reduce cross-machine incompatibilities.  So probably the simplest arrangement long-term would be to batch up such changes on a branch (next-major-release or something) so that people who want them can at least compile from that branch.  We could also consider making releases on both branches.

Adding features (preference file switches) that are relevant only to the client is easy -- no need for major version-number changes.

I can help with OS X testing, and really there is not much difference between OS X and linux development.  Windows has always been a problem and will continue to be...

**Comment by mintsoft** [comment]
I looked at adding Windows build support into travis a while ago, I think I was close to getting it working. Either way I'm happy to provide testing at least for Windows

**Comment by glondu** [comment]
> make sure everything builds on all platforms

It (i.e. current master, 641c3f3) builds at least on Linux (with OCaml 4.08.1). `-selftest` works and synchronizing my own stuff with it also works. I've seen reports of users having successfully built it on OSX (don't know about the graphical interface though). No idea about Windows.

I believe we can release in the current state, and make a minor release if an issue is found on some platform.

> update the minor version number

I think a major version bump is needed this time (see recent discussion on unison-users). What about bumping to 2.52.0?

> make sure documentation is rebuilt correctly

This seems to work.

> either put new binaries on download page in Unison web site or else just make that page point to GitHub

It sounds more convenient to make it point to GitHub. But it's your decision.

**Comment by tleedjarv** [comment]
I can confirm building and basic tests work on:
illumos
macOS (GitHub Actions build)
FreeBSD
(and Linux, as already mentioned)
all with OCaml 4.08.1 and all just text UI.

**Comment by bcpierce00** [comment]
> make sure everything builds on all platforms
> 
> It (i.e. current master, 641c3f3 <https://github.com/bcpierce00/unison/commit/641c3f3052f5072c6af253fedbc21261687920ee>) builds at least on Linux (with OCaml 4.08.1). -selftest works and synchronizing my own stuff with it also works. I've seen reports of users having successfully built it on OSX (don't know about the graphical interface though). No idea about Windows.
> 
> I believe we can release in the current state, and make a minor release if an issue is found on some platform.
> 
OK.
> update the minor version number
> 
> I think a major version bump is needed this time (see recent discussion on unison-users). What about bumping to 2.52.0?
> 
I’m always sad when we bump the major version number.  

> make sure documentation is rebuilt correctly
> 
> This seems to work.
> 
> either put new binaries on download page in Unison web site or else just make that page point to GitHub
> 
> It sounds more convenient to make it point to GitHub. But it's your decision.
> 
Github!

   - B

**Comment by glondu** [comment]
I've prepared the new release in https://github.com/bcpierce00/unison/pull/369 . I will merge it and tag in 2 weeks.

> make sure documentation is rebuilt correctly

I see what you mean now. I've regenerated the NEWS and strings.ml file. The `export` Makefile target is a bit ad hoc, though.

**Comment by bcpierce00** [comment]
> I've prepared the new release in #369 <https://github.com/bcpierce00/unison/pull/369> . I will merge it and tag in 2 weeks.
> 
Great.  Since it will bump the major version number, we should try to get as many major-version-breaking changes in as possible, so we don’t have to do it again for a while.  Are there any such that are still pending discussion?
> make sure documentation is rebuilt correctly
> 
> I see what you mean now. I've regenerated the NEWS and strings.ml file. The export Makefile target is a bit ad hoc, though.
> 
That’s putting it mildly.  :-)

**Comment by tleedjarv** [comment]
If the major version is going to be bumped anyway, I have two larger major-version-breaking changes that I'd like to discuss (and some smaller non-breaking changes as well).
The two would be #359 and #368. Both are actually working but require a solution review from someone more familiar with Unison architecture. Plus, I have an open question on both of them. @bcpierce00 do you think you will find time for that, or are there any other Unison experts here I could ping?

**Comment by bcpierce00** [comment]
I will take a look.

At the same time, can we step back and review which changes are forcing the version number bump?  Unless it is going to create an unmanageable degree of overhead, there might be a lot of value in maintaining both 2.51 and 2.52 branches.

**Comment by tleedjarv** [comment]
I walked through the changes since v2.51.2 tag. I didn't spot anything obviously incompatible but two changes caught my eye.

First, this one:
https://github.com/bcpierce00/unison/commit/ee8267b310d3b5ee51ccb16de2776167e50eb9d3#diff-7ff68864b19e3a361d5ec33e255d29eb
But I can't see how that would cause incompatibility.

Then:
https://github.com/bcpierce00/unison/commit/4a74472df70ebfe7e1b62c41e0171f3f532ec302#diff-b3da98720125ecd8bc66ff8e53761f3f
Here, the semantics have changed. This hash function is then used by `Update.checkArchive` at https://github.com/bcpierce00/unison/blob/master/src/update.ml#L221-L264, which in turn is used by `Update.commitUpdates` at https://github.com/bcpierce00/unison/blob/master/src/update.ml#L2113-L2147 to verify that archive checksums match.

This would trigger the incompatibility with case insensitive mode when Unicode is not used and there are non-ASCII characters.

**Comment by glondu** [comment]
> I walked through the changes since v2.51.2 tag.

Me too. The only changes that caught my eye are the changes from `Sort.list (<)` to `List.sort compare`. I am not sure they are equivalent (especially if there are equalities).

> First, this one:
> [ee8267b#diff-7ff68864b19e3a361d5ec33e255d29eb](https://github.com/bcpierce00/unison/commit/ee8267b310d3b5ee51ccb16de2776167e50eb9d3#diff-7ff68864b19e3a361d5ec33e255d29eb)
> But I can't see how that would cause incompatibility.

Me neither.

> Then:
> [4a74472#diff-b3da98720125ecd8bc66ff8e53761f3f](https://github.com/bcpierce00/unison/commit/4a74472df70ebfe7e1b62c41e0171f3f532ec302#diff-b3da98720125ecd8bc66ff8e53761f3f)
> Here, the semantics have changed. This hash function is then used by `Update.checkArchive` at https://github.com/bcpierce00/unison/blob/master/src/update.ml#L221-L264, which in turn is used by `Update.commitUpdates` at https://github.com/bcpierce00/unison/blob/master/src/update.ml#L2113-L2147 to verify that archive checksums match.
> 
> This would trigger the incompatibility with case insensitive mode when Unicode is not used and there are non-ASCII characters.

I believe lowercase_ascii is the same (implementation) as lowercase, they just wanted to make it obvious that Unicode was not handled with the new name.

**Comment by tleedjarv** [comment]
> > I walked through the changes since v2.51.2 tag.
> 
> Me too. The only changes that caught my eye are the changes from `Sort.list (<)` to `List.sort compare`. I am not sure they are equivalent (especially if there are equalities).

You are right, this may introduce a difference but I think the results of this function never reach the archive.

> I believe lowercase_ascii is the same (implementation) as lowercase, they just wanted to make it obvious that Unicode was not handled with the new name.

There must be a real difference, as the lowercase function was defined to work on Latin-1. But that's easy enough to check... let's head over to OCaml sources :) Indeed, this confirms there is a difference:
https://github.com/ocaml/ocaml/blob/trunk/stdlib/char.ml#L49-L69

I was just preparing to run some simple practical tests to see if this caused the incompatibility or not.

**Comment by tleedjarv** [comment]
I agree with @glondu on release management. My PRs are definitely not release-ready as-is, so that would expectedly postpone the release quite some time.
If major version bump can be avoided now then it is best to put out a release as soon as possible.

I do recommend merging #349 before the release, though (with Windows build disabled if that causes problems; and OCaml versions carefully selected). First, because it does not impact the code and the release itself. And second, then it is already present in the unison-2.52 branch for future bugfixes and minor releases. Perhaps travis build is sufficient enough to that end.

**Comment by tleedjarv** [comment]
> > I believe lowercase_ascii is the same (implementation) as lowercase, they just wanted to make it obvious that Unicode was not handled with the new name.
> 
> There must be a real difference, as the lowercase function was defined to work on Latin-1. But that's easy enough to check... let's head over to OCaml sources :) Indeed, this confirms there is a difference:
> https://github.com/ocaml/ocaml/blob/trunk/stdlib/char.ml#L49-L69
> 
> I was just preparing to run some simple practical tests to see if this caused the incompatibility or not.

I have done some testing and can now confirm that this is at least one of the incompatible changes. Maybe the easiest fix for 2.52 branch is to just revert the _ascii change?

**Comment by glondu** [comment]
> > I believe lowercase_ascii is the same (implementation) as lowercase, they just wanted to make it obvious that Unicode was not handled with the new name.
> 
> There must be a real difference, as the lowercase function was defined to work on Latin-1. But that's easy enough to check... let's head over to OCaml sources :) Indeed, this confirms there is a difference:
> https://github.com/ocaml/ocaml/blob/trunk/stdlib/char.ml#L49-L69

Oh. My bad.

> I have done some testing and can now confirm that this is at least one of the incompatible changes. Maybe the easiest fix for 2.52 branch is to just revert the _ascii change?

You mean revert and stay in the 2.51 series? (2.52 does not exist yet)

To the best of my knowledge, the last release in the 2.51 series is 2.51.2 and there is no "official" unison-2.51 branch. In my fork, I've [created one](https://github.com/glondu/unison/tree/unison-2.51) that branches off the v2.51.2 tag. It contains minimal changes to make it compile with OCaml 4.08.1 (but does not compile with OCaml 4.10.1 as it requires safe strings, which involves quite invasive changes).

So the last release is 2.51.2, made on Jan 27, 2018. Is it really an issue to release a new version now, 2.5 years later, that is incompatible? My opinion is no. I don't think we should strive to make a 2.51.2-compatible release based on master now. Making a release, say 2.51.3, based on my unison-2.51 branch (and making it "official") might make sense, though.

@bcpierce00 What do you think?

**Comment by glondu** [comment]
> I do recommend merging #349 before the release, though (with Windows build disabled if that causes problems; and OCaml versions carefully selected). First, because it does not impact the code and the release itself. And second, then it is already present in the unison-2.52 branch for future bugfixes and minor releases. Perhaps travis build is sufficient enough to that end.

OK, I've just merged it.

**Comment by tleedjarv** [comment]
> You mean revert and stay in the 2.51 series? (2.52 does not exist yet)

Ah, no, I messed up the versions. I meant that a 2.51.2-compatible release could be made from current master with just that one change reverted (don't need to revert the entire commit).

**Comment by bcpierce00** [comment]
> So the last release is 2.51.2, made on Jan 27, 2018. Is it really an issue to release a new version now, 2.5 years later, that is incompatible? My opinion is no. I don't think we should strive to make a 2.51.2-compatible release based on master now. Making a release, say 2.51.3, based on my unison-2.51 branch (and making it "official") might make sense, though.

> @bcpierce00 What do you think?

I think you may be surprised how much unhappiness a major version bump causes.  However, as has been pointed out in this thread, Unison is not a big project at the moment: the most precious resource is developer energy.  So, if going to 2.52 is the easiest way forward, I won't object.

**Comment by bcpierce00** [comment]
> You mean revert and stay in the 2.51 series? (2.52 does not exist yet)
> 
> Ah, no, I messed up the versions. I meant that a 2.51.2-compatible release could be made from current master with just that one change reverted (don't need to revert the entire commit).
> 
… On the other hand, if staying 2.51-compatible is as easy as reverting once small change, then that seems clearly better.

**Comment by bcpierce00** [comment]
And:

> On another side, but still related to release management, I propose the following: after the release of 2.52.0, bump the major version number in master (so to 2.53) and start merging new features. After, say, 6 months of inactivity in master, release 2.53.0. In parallel, maintain a unison-2.52 branch with only (compatible) bugfixes, and make 2.52.x releases after, say, 3 months of inactivity in that branch.

> I believe the wish to make new feature-ful releases compatible with the previous version is vain. First, because of this big OCaml-induced incompatibility (which I am working on, by the way). Second, in my humble opinion, the Unison project is not big / active enough to afford this constraint.

If we can get into a state where adding features (or changing OCaml compilers!) doesn't introduce such annoying incompatibilities, that would be great.  For the moment, the proposed management strategy sounds fine.

**Comment by gdt** [comment]
So now we have a master that is compatible, and it seems we are close to being able to release.   I think other than deciding about merging the windows build improvements, we're almost there.

**Comment by glondu** [comment]
I think the windows build improvements (#380) would be valuable to have in the 2.51.3 release, especially since they should not affect the other builds (except for this GitHub Action stuff which I don't really understand).

**Comment by hmeine** [comment]
I am not sure what's the right place to report this, but I just git-bisected a build problem on Ubuntu 18.04 which comes with OCaml 4.05.0, and found that @glondu's change in 4a74472df70ebfe7e1b62c41e0171f3f532ec302 leads to

    ocamlopt -g -unsafe-string -I lwt -I ubase -I system -I fsmonitor -I fsmonitor/linux -I fsmonitor/windows -I system/generic -I lwt/generic -c /home/hmeine/unison/src/lwt/generic/lwt_unix_impl.ml
    File "/home/hmeine/unison/src/lwt/generic/lwt_unix_impl.ml", line 340, characters 16-33:
    Error: Unbound module Stdlib
    Makefile.OCaml:411: recipe for target 'lwt/generic/lwt_unix_impl.cmx' failed

with that system / ocaml version.  (I usually use pre-built binaries, but I was trying to update from 2.48.4.)

**Comment by glondu** [comment]
Yes, the use of Stdlib indeed raises the minimum needed OCaml version to 4.07.0. In your case, I would suggest you to install a more recent version of ocaml (using e.g. opam) before building unison.

**Comment by gdt** [comment]
> I am not sure what's the right place to report this, but I just git-bisected a build problem on Ubuntu 18.04 which comes with OCaml 4.05.0, and found that @glondu's change in [4a74472](https://github.com/bcpierce00/unison/commit/4a74472df70ebfe7e1b62c41e0171f3f532ec302) leads to
> 
> ```
> ocamlopt -g -unsafe-string -I lwt -I ubase -I system -I fsmonitor -I fsmonitor/linux -I fsmonitor/windows -I system/generic -I lwt/generic -c /home/hmeine/unison/src/lwt/generic/lwt_unix_impl.ml
> File "/home/hmeine/unison/src/lwt/generic/lwt_unix_impl.ml", line 340, characters 16-33:
> Error: Unbound module Stdlib
> Makefile.OCaml:411: recipe for target 'lwt/generic/lwt_unix_impl.cmx' failed
> ```
> 
> with that system / ocaml version. (I usually use pre-built binaries, but I was trying to update from 2.48.4.)

This is not the right place!   It belongs in a new issue.   But the big question is if you have an ocaml version that is compliant with the requirements documented in the sources.  If so, it's a bug and please open an issue.  If not, then you can file a bug that the min version should not have been raised, if you think that's really the wrong thing for the project, or just update ocaml, if you think it's ok.    I'll refrain from further comments in this issue.

**Comment by hmeine** [comment]
I thought it may be relevant to the release (and therefore this issue), because it might not be expected from a patch release that the version requirements change.

Maybe the release should bump the minor version then?

**Comment by gdt** [comment]
Please open a new issue.   I see the point that it might be relevant, but mentioning it here is sufficient.  WIthout a separate issue, there is no reasonable way to discuss and track this.

**Comment by rivy** [comment]
@gdt @glondu @bcpierce00 , I've gotten a fully automated GHA CICD workflow working for all platforms (#380). The windows builds are working, but I need some help refactoring the windows code changes. Notably, windows testing is *not* working (and is currently disabled), although personal testing of synchronization using the windows executable is successful. I can't actually find *any* windows build that completes `unison -selftest`.

Any help would be appreciated.

**Comment by bcpierce00** [comment]
I’m not certain that there was ever a build on Windows that could run the tests.  I would just ignore that and forge ahead.

**Comment by gdt** [comment]
v2.51.3_rc2 has been tagged, and will likely be released this weekend absent adverse reports

**Comment by gdt** [comment]
v2.51.3_rc2 has been tagged, and will likely be released this week absent adverse reports.
(The release is within measurable distance of the end.)

**Comment by gdt** [comment]
v2.51.3 has been released.   @lilyball  thank you for your patience!
